### ISSUE TITLE
AWS Auto Scaling: Where to obtain Nginx Cert & Key?

### ISSUE THREAD

**Issue by Ir0nOx** [kasmtech/workspaces-issues#224]
To whom it may concern,

I have created a Kasm multi-server architecture and would like to take advantage of the cloud auto-scaling capabilities (I already have the license). However, when trying to establish  auto-scaling within Kasm, I'm not sure where to obtain the Nginx Cert and Key. The website is hosted by AWS's Route 53 and the SSL/TLS certs were created by AWS Certificate Manager (ACM).

As stated in the AWS Certificate Manager User Guide:
_"You cannot use ACM to install a public certificate directly on your AWS based website or application. You must use one of the services integrated with ACM."_

AWS does not provide the ability to obtain PEM certs in ACM. 

I could obtain the PEM hosted within the Kasm Database; however, I am not sure that would be a solution.

Your assistance is greatly appreciated. Thank you!

**Comment by wmkuipers** [comment]
Why not use a self signed certificate on the Kasm web instance, and add an ALB (as reverse proxy) in front of that, using the ACM issues certificate?

**Comment by mmcclaskey** [comment]
Kasm allows for two different architectures, one where all traffic flows through the API servers and one where the iframe for the desktop connection goes directly to the agent. In the 'direct to agent' architecture, a wild card cert for the domain is needed on the agent. In the default architecture, where all traffic flows through the API servers, you can just use any self signed cert. If you don't want to generate one yourself, just SSH to one of the agents and look under /opt/kasm/current/certs and you can copy the content of the nginx cert and key respectively.

**Comment by Ir0nOx** [comment]
@mmcclaskey 

Thank you for the quick response. I'll give that try!

**Comment by Ir0nOx** [comment]
Hi @mmcclaskey,

Does a launch template and auto scaling group need to be created? Also, I have yet to test, but any reported issues using Elastic File Storage (EFS) with Kasm Auto scaling? I'm guessing inputting a bootstrap script in the "Startup Script" should suffice.

**Comment by mmcclaskey** [comment]
If you are referring to AWS EC2 auto scaling there is no need to configure anything, Kasm handles auto scaling itself. 

We do have clients that use EFS for persistent profile storage and it works fine. You will need to modify the startup script to install and configure that on the host. I believe you will also need to ensure the EC2 IAM profile has the proper policy to access EFS.

**Comment by Ir0nOx** [comment]
Yes. I'm sorry for not clarifying that I was referring to AWS EC2 auto scaling. Thank you @mmcclaskey. 

Any future implementations regarding **not** having to assign Kasm an AWS Access and Secret keys for auto scaling? AWS roles avoid having to hardcode access keys for applications.

**Comment by mmcclaskey** [comment]
Nothing in our pipeline at this moment. We commonly will put the API/Manager servers in AWS and compute in Digital ocean or Oracle, just for example. So we can't assume that everyone auto scaling agents in AWS is running their API/Managers in AWS. Also there are lots of people with hybrid architectures, running baseline demand on-premise and auto-scaling to AWS to support peak demand. Correct me if I'm wrong, but I believe programmatic access is the only way to universally support this across any deployment configuration.

**Comment by Ir0nOx** [comment]
Hello @mmcclaskey & @j-travis,

Kasm's auto-scaling capabilities worked well. It was a bit of a pain to generate a proper and secure IAM policy to assign; there was a lot of error reading in Kasm to discover the hindrances. Anyways, it is functional. However, I am unable to discard the cloud auto scaling agents that errored (shown below): 

![image](https://user-images.githubusercontent.com/108290130/182403967-366aabb4-8d22-4b63-afda-d61478667e3c.png)

It has been in "delete_pending" status for 29 days. (UPDATE)

**Comment by GadgetGeekNI** [comment]
Do you have an example of that IAM policy? Would be really helpful to see! 

I have been working on one but I think I need to tweak it more. Unsure what I'm missing but it still doesn't play nicely.. (Created with Terraform)

data "aws_iam_policy_document" "kasm_agent_autoscaling_iam_policy" {

  #checkov:skip=CKV_AWS_111:The policy needs to be able to create keys and destroy autoscaled instances.
  statement {
    sid = "KasmAutoScaleGet"
    actions = [
      "ec2:DescribeInstances",
      "ec2:DescribeKeyPairs",
      "ec2:DescribeNetworkInterfaces",
      "ec2:CreateTags",
      "ec2:DeleteTags",
      "ec2:TerminateInstances"
    ]
    resources = ["arn:aws:ec2:${var.region}:${data.aws_caller_identity.current.id}:*"]
  }

  statement {
    sid = "AutoScaleKasmAgents"
    actions = [
      "ec2:RunInstances",
      "ec2:CreateTags",
      "ec2:DeleteTags",
      "ec2:CreateKeyPair",
      "ec2:ImportKeyPair"
    ]
    resources = [
      module.vpc.private_subnet_arns[2],
      aws_security_group.kasm_agent_sg.arn,
      aws_ami_copy.ubuntu_20_04_encrypted_ami.arn,
      "arn:aws:ec2:${var.region}:${data.aws_caller_identity.current.id}:network-interface/*",
      "arn:aws:ec2:${var.region}:${data.aws_caller_identity.current.id}:instance/*",
      "arn:aws:ec2:${var.region}:${data.aws_caller_identity.current.id}:volume/*",
      "arn:aws:ec2:${var.region}:${data.aws_caller_identity.current.id}:key-pair/*"
    ]
  }
  statement {
    sid = "KMSEncryptionForEC2"
    actions = [
      "kms:DescribeKey",
      "kms:Decrypt",
      "kms:Encrypt",
      "kms:CreateGrant"
    ]
    resources = [aws_kms_key.key_services[0].arn]
  }
  statement {
    sid       = "AssignInstanceProfile"
    actions   = ["iam:PassRole"]
    resources = [aws_iam_role.kasm_ssm_ec2_agent_session_profile.arn]
  }
}

**Comment by teja-kasm** [comment]
We’re currently enhancing our GitHub issue tracker to improve organization and efficiency.

As part of this process, we’re closing older issues to help keep the tracker focused on current and active topics.

If this issue is still relevant or you’re still experiencing the problem, please feel free to create a new issue.

Thank you for your understanding!
